components/function-controller/pkg/apis/serverless/v1alpha1/function_validation.go

```diff
@@ -45,64 +44,81 @@ type ValidationConfig struct {
 	Function     MinFunctionValues
 	BuildJob     MinBuildJobValues
 }
+type validationFunction func(*ValidationConfig) error
 
-func (fn *Function) performBasicValidation(ctx context.Context) *apis.FieldError {
-	return fn.validateObjectMeta(ctx).Also(
-		fn.Spec.validateSource(),
-		fn.Spec.validateEnv(ctx),
-		fn.Spec.validateLabels(),
-		fn.Spec.validateReplicas(ctx),
-		fn.Spec.validateFunctionResources(ctx),
-		fn.Spec.validateBuildResources(ctx),
-	)
+func (fn *Function) performBasicValidation(vc *ValidationConfig) error {
```

Comment:
What do you think about returning from this method basicValidationFns, because this function is always called to get validationFns.

I like the idea of validationFns array as you implemented it.